common/txmgr/confirmer.go

```diff
@@ -1004,22 +1024,30 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) han
 	}
 }
 
-// EnsureConfirmedTransactionsInLongestChain finds all confirmed txes up to the depth
+// EnsureConfirmedTransactionsInLongestChain finds all confirmed txes up to the earliest head
 // of the given chain and ensures that every one has a receipt with a block hash that is
 // in the given chain.
 //
 // If any of the confirmed transactions does not have a receipt in the chain, it has been
 // re-org'd out and will be rebroadcast.
-func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) EnsureConfirmedTransactionsInLongestChain(ctx context.Context, head types.Head[BLOCK_HASH]) error {
-	if head.ChainLength() < ec.chainConfig.FinalityDepth() {
-		logArgs := []interface{}{
-			"chainLength", head.ChainLength(), "finalityDepth", ec.chainConfig.FinalityDepth(),
-		}
+func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) EnsureConfirmedTransactionsInLongestChain(ctx context.Context, head types.Head[BLOCK_HASH], latestFinalizedHeadNumber int64) error {
+	logArgs := []interface{}{
+		"chainLength", head.ChainLength(), "latestFinalizedHead number", latestFinalizedHeadNumber,
+	}
+
+	if head.BlockNumber() < latestFinalizedHeadNumber {
+		errMsg := "current head is shorter than latest finalized head"
+		ec.lggr.Errorw(errMsg, append(logArgs, "head block number", head.BlockNumber())...)
+		return errors.New(errMsg)
+	}
+
+	calculatedFinalityDepth := uint32(head.BlockNumber() - latestFinalizedHeadNumber)
+	if head.ChainLength() < calculatedFinalityDepth {
 		if ec.nConsecutiveBlocksChainTooShort > logAfterNConsecutiveBlocksChainTooShort {
-			warnMsg := "Chain length supplied for re-org detection was shorter than FinalityDepth. Re-org protection is not working properly. This could indicate a problem with the remote RPC endpoint, a compatibility issue with a particular blockchain, a bug with this particular blockchain, heads table being truncated too early, remote node out of sync, or something else. If this happens a lot please raise a bug with the Chainlink team including a log output sample and details of the chain and RPC endpoint you are using."
+			warnMsg := "Chain length supplied for re-org detection was shorter than actual finality depth. Re-org protection is not working properly. This could indicate a problem with the remote RPC endpoint, a compatibility issue with a particular blockchain, a bug with this particular blockchain, heads table being truncated too early, remote node out of sync, or something else. If this happens a lot please raise a bug with the Chainlink team including a log output sample and details of the chain and RPC endpoint you are using."
```

Comment:
What's the difference between `FinalityDepth` and `actual finality depth` ?

Reply:
`FinalityDepth` written this way usually refers to our EVM config but `actual finality depth` refers to the depth we calculate from the latest head to the finalized head. Your question makes me consider if "actual" is descriptive enough though. Maybe it should be `calculated finality depth` or `effective finality depth`?

Reply:
Or maybe even more generally something like `Chain length supplied for re-org detection was shorter than the depth from the latest head to the finalized head`

Reply:
Yeah that's better imo, the extra description is helpful

Reply:
`FinalityDepth` was hardcoded I believe, actual finality depth is what we calculated using the latestFinalizedBlockHead (from head tracker) and current head